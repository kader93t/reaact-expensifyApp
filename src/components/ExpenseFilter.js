import React from 'react'
import { connect } from 'react-redux';
import {setEndDate, setStartDate, setTextFilter} from "./../actions/filter"
import { setSortByDate, setSortByAmount } from "./../actions/filter"
import { DateRangePicker } from "react-dates"


export class ExpensesFilters extends React.Component{
    state = {
        focusedInput: undefined
    }

    setSortByDate = () => {
        this.props.setSortByDate();
    }

    setSortByAmount = () => {
        this.props.setSortByAmount();
    }

    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    }

    setTextFilter = (e) => {
        this.props.setTextFilter(e.target.value);
    }

    onSortChange = (e) => {      
        switch (e.target.value) {
            case "date": {
                this.setSortByDate();
                break;
            }
            case "amount":{
                this.setSortByAmount();
                break;
            }
        }
    }
    
    render() {
        
        return  ( <div>        
            
            <input type="text" placeholder="Text filter" value={this.props.filter.text} onChange={this.setTextFilter}/>
            
            <select
                value={this.props.filter.sortBy}
                onChange={this.onSortChange}>
                <option value="amount">Amount</option>
                <option value="date">Date</option>
            
            </select>
            
            <DateRangePicker
                startDate={this.props.filter.startDate}
                startDateId="startDate"
                endDate={this.props.filter.endDate}
                endDateId="endDate"
                onDatesChange={
                    this.onDatesChange
                }
                focusedInput={this.state.focusedInput} // PropTypes.oneOf([START_DATE, END_DATE]) or null,
                onFocusChange={focusedInput => this.setState({ focusedInput })} // PropTypes.func.isRequired,
                isOutsideRange={() => false}
                showClearDates={true}
                numberOfMonths = {1}
        />
            </div>)
            }
}

const mapStateToProps = (state) => {
    return {
        filter: state.filter    
    }
};

const mapDispatchToProps = (dispatch) => ({
    setEndDate: (endDate) => dispatch(setEndDate(endDate)),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setSortByAmount: () => dispatch(setSortByAmount()),
    setSortByDate: () => dispatch(setSortByDate()),
    setTextFilter: (text) => dispatch(setTextFilter(text))
});
export default connect(mapStateToProps, mapDispatchToProps)(ExpensesFilters);

