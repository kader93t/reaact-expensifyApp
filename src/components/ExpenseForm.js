import React, { Component } from "react";
import { SingleDatePicker } from "react-dates";

import moment from "moment";

export default class ExpenseForm extends Component {
  constructor(props) {
    super(props);
    this.state = {
      description: props.expense ? props.expense.description : "",
      note: props.expense ? props.expense.note : "",
      amount: props.expense ? props.expense.amount.toString() : "",
      createAt: props.expense ? moment(props.expense.createAt) : moment(),
      calendarFocused: false,
      error: "",
    };
  }

  onChangeDescription = (e) => {
    const description = e.target.value;
    this.setState((oldState) => ({
      ...oldState,
      description,
    }));
  };

  onChangeNote = (e) => {
    const note = e.target.value;
    this.setState((oldState) => ({
      ...oldState,
      note,
    }));
  };

  onChangeAmount = (e) => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,3})?$/)) {
      this.setState((oldState) => ({
        ...oldState,
        amount,
      }));
    }
  };

  onSubmit = (e) => {
    e.preventDefault();
    if (!this.state.description || !this.state.amount) {
      this.setState((oldState) => ({
        error: "One of the important field is empty",
      }));
    } else {
      this.setState((oldState) => ({
        error: "",
      }));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount),
        createAt: this.state.createAt.valueOf(),
        note: this.state.note,
      });
    }
  };

  render() {
    return (
      <form className="form" onSubmit={this.onSubmit}>
        {this.state.error && <p className="form__error">{this.state.error}</p>}
        <input
          className="text-input"
          type="text"
          placeholder="Description"
          autoFocus
          value={this.state.description}
          onChange={this.onChangeDescription}
        />
        <input
          className="text-input"
          type="text"
          placeholder="Amount"
          value={this.state.amount}
          onChange={this.onChangeAmount}
        />

        <textarea
          className="input-area"
          placeholder="Add a note for your expense (optional)"
          value={this.state.note}
          onChange={this.onChangeNote}
        ></textarea>

        <SingleDatePicker
          date={this.state.createAt}
          onDateChange={(createAt) => {
            if (createAt) this.setState(() => ({ createAt }));
          }}
          focused={this.state.calendarFocused}
          onFocusChange={({ focused: calendarFocused }) => {
            this.setState(() => ({ calendarFocused }));
          }}
          id="my_id"
          isOutsideRange={() => false}
          numberOfMonths={1}
        />
        <div >
          <button className='button'> Add Expnese </button>
        </div>
      </form>
    );
  }
}
