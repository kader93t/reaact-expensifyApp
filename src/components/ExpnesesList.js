import React from 'react'

import { connect } from 'react-redux'
import ExpensItem from "./ExpenseItem"
import selectExpenses from "./../selectors/expenses"

export const ExpensesList = (props) => (
    <div>
        <h1>Expenses List</h1>
        {props.expenses.map((expense) => (
          
            <ExpensItem key={expense.id} {...expense}/>
        ))}
    </div>
);

const mapStateToProps = (state) => (
    {
        expenses: selectExpenses(state.expenses,state.filter)
    }
);

export default connect(mapStateToProps)(ExpensesList);
