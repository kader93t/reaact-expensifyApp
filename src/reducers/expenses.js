// Expenses Reducer

const expenseDefaultState = [];
export default (state = expenseDefaultState, action) => {
    
    switch (action.type) {
        case "ADD_EXPENSE":
            return [       
                ...state, action.expenses
            ]
        case "REMOVE_EXPENSE":
            return state.filter((expense) => {
                return expense.id !== action.id;
            })
        case "UPDATE_EXPENSE":
            return state.map((expense) => {
                if (expense.id === action.id) {
                    return {
                        ...expense,
                        ...action.updates
                    }
                } else {
                    return expense;
                }
            });
            ;
        default:
            return state;
    }
}


