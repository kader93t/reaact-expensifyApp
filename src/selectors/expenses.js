
export default (expenses, { startDate, endDate, text, sortBy }) => {
    const visibleExpenses = expenses.filter((expense) => {
        const matchStartDate = !startDate || startDate.isSameOrBefore(expense.createAt, "day");
        const matchEndDate = !endDate || endDate.isSameOrAfter(expense.createAt, "day");
        const matchText = !text || expense.description.toLowerCase().includes(text.toLowerCase());
        return matchEndDate && matchStartDate && matchText;
    });
    return visibleExpenses.sort((a, b) => {
        if (sortBy === "date")
            return a.createAt < b.createAt ? 1 : -1;
        else if (sortBy === "amount")
            return a.amount < b.amount ? 1 : -1;
    });
};
