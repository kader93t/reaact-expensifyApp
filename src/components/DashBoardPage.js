import React from "react"
import ExpensesList from './ExpnesesList'
import ExpenseFilter from "./ExpenseFilter"

const ExpenseDashBoardPage = () => (
    <div>
        Welcom to react Router
        <ExpenseFilter />
        <ExpensesList />
    </div>
);


export default ExpenseDashBoardPage;