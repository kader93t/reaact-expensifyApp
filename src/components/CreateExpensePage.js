import React, { Component } from 'react'
import { connect } from 'react-redux'
import { addExpense }from "./../actions/expneses.js"
import ExpenseForm from "./ExpenseForm"

export class CreateExpensePage extends Component {

    onSubmit = (expense) => {
        this.props.onSubmit(expense);
        this.props.history.push("/");
    };

    render() {
        return (
            <div>
                <h1>Add expense </h1>
                <ExpenseForm onSubmit={this.onSubmit} />
            </div>
        );
    }
}

const mapDispatchToProps = (dispatch) => (
    {
        onSubmit: (expense) => dispatch(addExpense(expense))
    }
);

export default connect(undefined, mapDispatchToProps)(CreateExpensePage);
