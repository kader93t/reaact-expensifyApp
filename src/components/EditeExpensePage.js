import React from "react"
import { connect } from "react-redux";
import { startEditeExpense, startRemoveExpense } from "../actions/expneses";
import ExpenseForm from "./ExpenseForm"

export class EditeExpensePage extends React.Component {
    
    onSubmit = (expense) => {
        this.props.modifieExpense(this.props.expense.id, expense);
        this.props.history.push("/");
    }
    
    onButtonClick = (e) => {
        this.props.onButtonClick(this.props.expense.id);
        this.props.history.push("/");
    }
    
    render() {
        return (<div>
            <ExpenseForm expense={this.props.expense}
                onSubmit={this.onSubmit}
            />
            <button
                onClick={this.onButtonClick}
            > Remove </button>
        </div>
        );
    }
}

const mapStateToProps = (state, props) => 
    ({
    expense: state.expenses.find((expense) => expense.id === props.match.params.id)
}
    );

const mapDispatchToProps = (dispatch) => (
    {
        modifieExpense: (id, expense) => dispatch(startEditeExpense(id, expense)),
        onButtonClick: (id)=> dispatch(startRemoveExpense(id))
    }
)

export default connect(mapStateToProps,mapDispatchToProps)(EditeExpensePage);