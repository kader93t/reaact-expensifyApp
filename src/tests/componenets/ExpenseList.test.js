import React from 'react'
import { shallow } from "enzyme"
import { ExpensesList } from './../../components/ExpnesesList';
import { expenses } from "./../fixture/expenses"
test("should render the expenses list ", () => {
    const wrapper = shallow(<ExpensesList expenses={expenses} />);
    expect(wrapper).toMatchSnapshot();

});

test("should render empty list", () => {
    const wrapper = shallow(<ExpensesList expenses={[]} />);
    expect(wrapper).toMatchSnapshot();
})