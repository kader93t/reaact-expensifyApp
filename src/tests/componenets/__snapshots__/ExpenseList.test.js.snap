// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render empty list 1`] = `
<div>
  <h1>
    Expenses List
  </h1>
</div>
`;

exports[`should render the expenses list  1`] = `
<div>
  <h1>
    Expenses List
  </h1>
  <ExpenseItem
    amount={2.12}
    createAt={0}
    description="hello1"
    id={1}
    key="1"
    note="yesterday1"
  />
  <ExpenseItem
    amount={2.12}
    createAt={0}
    description="hello2"
    id={2}
    key="2"
    note="yesterday2"
  />
  <ExpenseItem
    amount={2.12}
    createAt={0}
    description="hello3"
    id={3}
    key="3"
    note="yesterday3"
  />
</div>
`;
